### ISSUE TITLE
Better docs for image substitutions

### ISSUE THREAD

**Issue by ensonic** [bazelbuild/rules_k8s#68]
Lets assume I have a deployment:
```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: web-frontend
spec:
  replicas: 1
  template:
    spec:
      - name: web-frontend
        image: XXXX
```
and a build rule:
```
k8s_object(
  name = "x-deployment",
  kind = "deployment",
  template = ":deployment.yaml",
  images = {
    "XXXX": "//src/docker/web_frontend:web-frontend",
  }
)
```

I get:
```
INFO: Running command line: bazel-bin/src/kubernetes/web-frontend/x-deployment
Traceback (most recent call last):
  File "/usr/local/google/home/ensonic/.cache/bazel/_bazel_ensonic/execroot/__main__/bazel-out/local-fastbuild/bin/src/kubernetes/web-frontend/x-deployment.runfiles/__main__/../io_bazel_rules_k8s/k8s/resolver.py", line 180, in <module>
    main()
  File "/usr/local/google/home/ensonic/.cache/bazel/_bazel_ensonic/execroot/__main__/bazel-out/local-fastbuild/bin/src/kubernetes/web-frontend/x-deployment.runfiles/__main__/../io_bazel_rules_k8s/k8s/resolver.py", line 166, in main
    (tag, digest) = Publish(transport, args.image_chroot, **kwargs)
  File "/usr/local/google/home/ensonic/.cache/bazel/_bazel_ensonic/execroot/__main__/bazel-out/local-fastbuild/bin/src/kubernetes/web-frontend/x-deployment.runfiles/__main__/../io_bazel_rules_k8s/k8s/resolver.py", line 136, in Publish
    name_to_replace = docker_name.Tag(name)
  File "/usr/local/google/home/ensonic/.cache/bazel/_bazel_ensonic/execroot/__main__/bazel-out/local-fastbuild/bin/src/kubernetes/web-frontend/x-deployment.runfiles/containerregistry/client/docker_name_.py", line 196, in __init__
    _check_tag(self._tag)
  File "/usr/local/google/home/ensonic/.cache/bazel/_bazel_ensonic/execroot/__main__/bazel-out/local-fastbuild/bin/src/kubernetes/web-frontend/x-deployment.runfiles/containerregistry/client/docker_name_.py", line 80, in _check_tag
    _check_element('tag', tag, _TAG_CHARS, 1, 127)
  File "/usr/local/google/home/ensonic/.cache/bazel/_bazel_ensonic/execroot/__main__/bazel-out/local-fastbuild/bin/src/kubernetes/web-frontend/x-deployment.runfiles/containerregistry/client/docker_name_.py", line 64, in _check_element
    % (name, element, min_len))
containerregistry.client.docker_name_.BadNameException: Invalid tag: , must be at least 1 characters
ERROR: Non-zero return code '1' from command: Process exited with status 1
```

The docs need to explain what format the key in the images param in the BUILD file needs to be and how to specify this in the deployment.yaml.

I also tried `$(GCP_REGISTRY)/$(GCP_PROJECT)/web-frontend` where GCP_REGISTRY and GCP_PROJECT are defines in my bazelrc, but I don't want to hardcode this in the template (deployment.yaml).

It looks like it tries to parse the key. E.g. when I change "XXXX" to "XXXX:latest" I get 
"""
containerregistry.client.docker_name_.BadNameException: A Docker registry domain must be specified.
"""

If that is the case, I don't understand how I can avoid hard-coding the registry domains into the templates. I think it would be nicer to specify this in k8s_defaults() as a param.

If I put `gcr.io/<project-id>/web-fontend:latest` into the deployment.yaml and `"gcr.io/<project-id>/web-fontend:latest": "//src/docker/web_frontend:web-frontend"` into the BUILD file, the rule runs, but it apparently interacts with the cloud (I need to have a valid regsitry login), I though it would only resolve (aka substitue) the template. Also the produced output template is not expanded, it only prints the resolved template to stdout.

**Comment by mattmoor** [comment]
At least today `XXXX` has to be a well-formed and fully-qualified image name.  This means, there must be a domain part (e.g. `gcr.io`) and a tag (e.g. `:latest`).  We could conceivably relax this by disabling strictness in the `docker_name.Tag`, or allowing the keys to replace arbitrary strings.  If this sounds appealing let's fork another issue for that piece?

I think the crux of this is that you want this:
> I also tried $(GCP_REGISTRY)/$(GCP_PROJECT)/web-frontend ...

These rules currently support stamp variables (IIRC), but not make variables (which I find annoying without `vardef` for defaults).  That said, stamp variables in the key of `images={}` isn't super useful without a templating system with similar stamping support ([related](https://github.com/bazelbuild/rules_jsonnet/issues/29)).

> ... where GCP_REGISTRY and GCP_PROJECT are defines in my bazelrc

This is where `image_chroot` comes in.  It allows you to rewrite an embedded reference into a developer namespace.  It's a bit ugly with strict names, but you can insert a dummy name and aways chroot it.  We do this in the examples (in part) for integration testing these features.


Let me know if there is anything I didn't speak to.

**Comment by mattmoor** [comment]
So this is sort of an artifact of how this grew up, and it certainly isn't immovable :)

Essentially, we started from a point where the yaml templates basically had to be independently useful, since the only modifications we performed were tag => digest.  With chrooting, the game changes considerably.

Let me TAL to confirm that this is a simple change (ideally by making it), and then we can replace arbitrary strings with digests.
